UploadDaemon/UploadDaemon.cs

```diff
@@ -0,0 +1,168 @@
+        private readonly UploadTask uploadTask;
+        private readonly FileSystem fileSystem;
+
+        /// <summary>
+        /// Main entry point. Expects a single argument: the path to a directory that contains the trace files to upload.
+        /// </summary>
+        public static void Main(string[] args)
+        {
+            if (IsAlreadyRunning())
+            {
+                Console.WriteLine("Another instance is already running.");
+                NotifyRunningDaemon();
+                return;
+            }
+
+            Config config;
+            try
+            {
+                logger.Debug("Reading config from {configFile}", Config.ConfigFilePath);
+                config = Config.ReadFromCentralConfigFile();
+            }
+            catch (Exception e)
+            {
+                logger.Error(e, "Failed to read config file {configPath}", Config.ConfigFilePath);
+                Environment.Exit(1);
+                return;
+            }
+
+            logger.Info("Starting upload daemon v{uploaderVersion}", Assembly.GetExecutingAssembly().GetName().Version.ToString());
+            var uploader = new UploadDaemon(config);
+            uploader.RunOnce();
+            uploader.ScheduleRegularUploads();
+            uploader.WaitForNotifications();
+        }
+
+        private static bool IsAlreadyRunning()
+        {
+            Process current = Process.GetCurrentProcess();
+            // If this process is not running from the exe file, we allow it
+            // Not sure when this happens, though
+            if (Assembly.GetExecutingAssembly().Location.Replace("/", "\\") != current.MainModule.FileName)
+            {
+                return false;
+            }
+
+            return Process.GetProcessesByName(current.ProcessName).Where(process => process.Id != current.Id).Any();
+        }
+
+        public UploadDaemon(Config config)
+        {
+            fileSystem = new FileSystem();
+
+            this.config = config;
+            if (config.DisableSslValidation)
+            {
+                HttpClientUtils.DisableSslValidation();
+            }
+
+            uploadTask = new UploadTask(config, fileSystem, new UploadFactory());
+        }
+
+        /// <summary>
+        /// Runs the timer action once synchronously.
+        /// </summary>
+        public void RunOnce()
+        {
+            lock(SequentialUploadsLock)
+            {
+                uploadTask.Run();
+            }
+        }
+
+        /// <summary>
+        /// Schedule uploads on a regular intervall.
+        /// </summary>
+        private void ScheduleRegularUploads()
+        {
+            Timer timer = new Timer();
+            timer.Elapsed += (sender, args) => RunOnce();
+            timer.Interval = TimerIntervalInMilliseconds;
+            timer.Enabled = true;
+        }
+
+        /// <summary>
+        /// Waits for notifications from subsequent executions of the Daemon. Terminates on
```

Comment:
where is that handling for the cancel key? do we really need that?

Reply:
I removed this already, after we agreed we don't need it on the phone. Updated the comment now as well. Good catch!